### PR TITLE
Implement local authentication prompt screen

### LINKED ISSUE #81: [Feature] New user - Local authentication prompt screen
@talmahaj commented on [Sun Jan 19 2020](https://github.com/tari-project/wallet-ios/issues/100)

See [parent](https://github.com/tari-project/tari/issues/1217)

### PR DESCRIPTION
**Pull request by jigneshsanghani**
Closes #81

Changes:
 - Implement local auth setup fragment that supports biometric and PIN types
 - Use it during onboarding

TODO:
 - Decide what to do if auth fails (Either due to hardware unavailable or user denies auth). Currently, it just displays an alert and allows user to go ahead.

### CODE REVIEW

**Comment by kukabi** [comment]
When all security is disabled on the phone the auth warning dialog appears right at the start of the activity. It feels a bit sudden, maybe it could show when the user clicks Enable Your PIN (I know at the code level we know if the phone is secured or not, but still I think it's better UX if it shows at the click of the button, Karim could tell better tho).

**Comment by kukabi** [comment]
When all security is disabled and I click "Proceed" in the auth warning dialog, an animation shows a black screen and a PIN image shows up and fades out. There's no security enabled and I'm proceeding, so this sequel looks irrelevant.
no_security_proceed

**Comment by kukabi** [comment]
If I enable PIN/pattern/pass then click Enable Your PIN, then click the back button at the PIN/pattern/pass entry screen, then click Proceed in the auth warning dialog in the local auth screen, then the PIN animation gets displayed as in the above comment
security_cancel_enter_pass_and_go_back
 and I'm let into the application.

**Comment by kukabi** [comment]
And when I enter the pass/PIN/pattern correctly I'm taken back to the previous screen and the PIN animation in the dark screen gets played. Are we sure this is the intended behaviour? I think it would make more sense UX-wise to see the dark screen PIN view appear first, then to be taken to the PIN/pass/pattern screen, then I enter whatever, and I'm taken back and I see the dark PIN screen, it dissolves into the intial auth screen and the flow is completed, I'm taken to home screen. Is there a showcase or demo video for this flow?

**Comment by kukabi** [comment]
I've just tested with fingerprints, and you can repeat the comments above the same for the fingerprint flows. If there's any kind of security enabled on the phone I'm not sure if we allow the user to ever use the app without passing device security.

File: app/src/main/java/com/tari/android/wallet/ui/activity/onboarding/OnBoardingFlowActivity.kt [hunk 1]
```diff
@@ -99,21 +104,50 @@ class OnboardingFlowActivity : BaseActivity(), IntroductionFragment.Listener {
...
+            fragment.fadeOutAllViewAnimation()
+        }
+
+        supportFragmentManager
+            .beginTransaction()
+            .setCustomAnimations(R.anim.fade_in, R.anim.fade_out)
+            .replace(R.id.onboarding_frame_container, LocalAuthFragment())
+            .commit()
+
+        removeCurrentFragment(
```
Comment (kukabi) [inline app/src/main/java/com/tari/android/wallet/ui/activity/onboarding/OnBoardingFlowActivity.kt]: Does `replace` not remove the current fragment?

File: app/src/main/java/com/tari/android/wallet/ui/fragment/onboarding/LocalAuthFragment.kt [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+            authSuccessAuthTypeImageView.setImageResource(R.drawable.numpad)
+            authSuccessAuthTypeTitleTextView.text = authPin
+        }
+
+        if (authType == AuthType.None) {
+            authFailed()
+        }
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
```
Comment (kukabi) [inline app/src/main/java/com/tari/android/wallet/ui/fragment/onboarding/LocalAuthFragment.kt]: Please group lifecycle functions together - preferably at the top.

File: app/src/main/java/com/tari/android/wallet/ui/fragment/onboarding/LocalAuthFragment.kt [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+            })
+    }
+
+    override fun onDetach() {
+        super.onDetach()
+        listener = null
+    }
+
+    @OnClick(R.id.local_auth_btn_use_auth)
+    fun onAuthClick() {
```
Comment (kukabi) [inline app/src/main/java/com/tari/android/wallet/ui/fragment/onboarding/LocalAuthFragment.kt]: Let's make the id's and click action functions follow what's written on the button. Maybe more like `onEnablePINButtonClick`/`onEnableAuthButtonClick`, or change the text on the button to go "Authenticate".

File: app/src/main/java/com/tari/android/wallet/ui/fragment/onboarding/LocalAuthFragment.kt [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+            override fun onAnimationEnd(animation: Animator?) {
+                super.onAnimationEnd(animation)
+                startHomeActivity()
+            }
+        })
+        animatorSet.start()
+    }
+
+    private fun startHomeActivity() {
+        listener?.onAuthTypeSetSuccess()
```
Comment (kukabi) [inline app/src/main/java/com/tari/android/wallet/ui/fragment/onboarding/LocalAuthFragment.kt]: Not sure if the name reflects what's being done here. It's correct if we're setting the authentication type, but it's actually done at the OS settings level.

File: app/src/main/java/com/tari/android/wallet/ui/fragment/onboarding/LocalAuthFragment.kt [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+    }
+
+    private fun playStartUpAnim() {
```
Comment (kukabi) [inline app/src/main/java/com/tari/android/wallet/ui/fragment/onboarding/LocalAuthFragment.kt]: Let's layout the animation functions chronologically, so this one comes before the group animation function above?

**Comment by kukabi** [comment]
After the last commit: if I click "Enable your PIN" when there's no security enabled on the device, I see the PIN animation, and when it finishes it says "Auth Failed" and the only option is to exit. Yet when I exit and restart the app, I see the good old auth screen, and I'm able to get into the app, as devised. I think it makes more sense to not show the PIN anim if there's no security enabled, and just mimic the behavior of the old Auth screen. We don't block the user if there's no lock on the phone. We just warn, and let them choose if they want to use the app despite, or not.

**Comment by kukabi** [comment]
When security is enabled, I click the "Enable you PIN" button, I see the PIN animation, the animation finishes, then the screen goes back to the same screen, then I see the Android PIN/pass screen. Is this the devised flow? I asked above, is there a design spec video for this flow? If the PIN animation has to show, then to me it makes more sense to show the PIN image, then be directed to PIN entry, then come back and see the PIN image again, PIN image disappears and I either continue or get taken back to the "Enable your PIN" screen. And again, if the PIN is not entered (i.e. back button is clicked), I should only be warned and should be able to close the dialog and stay at the screen.

**Comment by kukabi** [comment]
If i interrupt the flow by exiting the app at "Enter your PIN" screen, I get taken to the old auth screen when I restart the app. I would expect it take it from the same screen - I was setting up (or being introduced to) security, and I would like to finish that first.

File: app/src/main/java/com/tari/android/wallet/ui/activity/onboarding/OnBoardingFlowActivity.kt [hunk 6]
```diff
@@ -71,14 +76,14 @@ class OnboardingFlowActivity : BaseActivity(), IntroductionFragment.Listener {
             supportFragmentManager
                 .beginTransaction()
                 .replace(
-                    R.id.onboarding_create_emoji_id_container,
+                    R.id.onboarding_frame_container2,
```
Comment (kukabi) [inline app/src/main/java/com/tari/android/wallet/ui/activity/onboarding/OnBoardingFlowActivity.kt]: What does 1 & 2 imply? Front/back, or is it a chronological ordering?

**Comment by jigneshsanghani** [comment]
Hi @kukabi, local auth implementation at 1:08 duration in this video. https://www.dropbox.com/s/8ipek9ww93urboz/Testnet%20Wallet%20FTUE.mp4?dl=0

Ticket is not super clear about what should be done if
- Auth fails
- User does not have auth supported device
- User does not have auth enabled

Currently, I have considered all of them in a single case auth fail, and I think it makes sense as well because we don't want user to use wallet if he does not pass auth. User should have at least 1 method of auth enabled and should pass that auth in order to use our app. Let me know your thoughts, thanks!

**Pull request merged**